File: mdz/pytorch/PointNet/3_deploy/modelzoo/PointNet/src/PointNet.cpp

#include <icraft-xrt/core/session.h>
#include <icraft-xrt/dev/host_device.h>
#include <icraft-xrt/dev/buyi_device.h>
#include <icraft-backends/buyibackend/buyibackend.h>
#include <icraft-backends/hostbackend/cuda/device.h>
#include <icraft-backends/hostbackend/backend.h>
#include <icraft-backends/hostbackend/utils.h>
#include <fstream>
#include <opencv2/opencv.hpp>
#include "icraft_utils.hpp"
#include "yaml-cpp/yaml.h"
using namespace icraft::xrt;
using namespace icraft::xir;



int main(int argc, char* argv[])
{
	YAML::Node config = YAML::LoadFile(argv[1]);
	// icraft模型部署相关参数配置
	auto imodel = config["imodel"];
	// 仿真上板的jrpath配置
	std::string folderPath = imodel["dir"].as<std::string>();  
	bool run_sim = imodel["sim"].as<bool>();
    bool cudamode = imodel["cudamode"].as<bool>();
	std::string targetFileName;
	std::string JSON_PATH = getJrPath(run_sim,folderPath, imodel["stage"].as<std::string>());
	std::regex rgx3(".json");
	std::string RAW_PATH = std::regex_replace(JSON_PATH, rgx3, ".raw");
	// URL配置
	std::string ip = imodel["ip"].as<std::string>();
	// 可视化配置
	bool show = imodel["show"].as<bool>();
	bool save = imodel["save"].as<bool>();

	// 加载network
	Network network = loadNetwork(JSON_PATH, RAW_PATH);
	//初始化netinfo
	NetInfo netinfo = NetInfo(network);
	// 打开device
	Device device = openDevice(run_sim, ip,netinfo.mmu || imodel["mmu"].as<bool>());
	// 初始化session
	Session session = initSession(run_sim, network, device, netinfo.mmu || imodel["mmu"].as<bool>(), imodel["speedmode"].as<bool>(), imodel["compressFtmp"].as<bool>());
	// 开启计时功能
	session.enableTimeProfile(true);
	// session执行前必须进行apply部署操作
	session.apply();

	// 数据集相关参数配置
	auto dataset = config["dataset"];
	std::string imgRoot = dataset["dir"].as<std::string>();
	std::string imgList = dataset["list"].as<std::string>();
	std::string names_path = dataset["names"].as<std::string>();
	std::string resRoot = dataset["res"].as<std::string>();
	checkDir(resRoot);
	auto LABELS = toVector(names_path);
	// 模型自身相关参数配置
	auto param = config["param"];
	int N_CLASS = param["number_of_class"].as<int>();
	// 统计输入数量
	int index = 0;
	auto namevector = toVector(imgList);
	int totalnum = namevector.size();
	for (auto name : namevector) {
		progress(index, totalnum);
		index++;
		std::string img_path = imgRoot + '/' + name;
		// read FTMP and convert to Tensor
		auto img_tensor = hostbackend::utils::Ftmp2Tensor(img_path, network.inputs()[0].tensorType());
		
		dmaInit(run_sim, netinfo.ImageMake_on,img_tensor, device);
		
		std::vector<Tensor> outputs = session.forward({ img_tensor });
		// check outputs output1 = [1,40],output2 = [1,3,3],output3 = [1,64,64]
		// for(auto output : outputs){
		// 	std::cout << output.dtype()->shape << std::endl;
		// }
		if(!run_sim) device.reset(1);
		// 计时
		#ifdef __linux__
		device.reset(1);
		calctime_detail(session);
		#endif
		
		// post process 寻找最大值对应下标
		auto host_tensor = outputs[0].to(HostDevice::MemRegion());	
		auto pred = (float*)host_tensor.data().cptr();
        auto max_prob_ptr = std::max_element(pred, pred + N_CLASS);
		int max_index = std::distance(pred, max_prob_ptr);
		std::cout <<"\nPRED ="<<LABELS[max_index] << std::endl;
		if (save) {
        std::string save_path = resRoot + '/' + name;
        std::regex rgx("\\.(?!.*\\.)");
        std::string RES_PATH = std::regex_replace(save_path, rgx, "_result.");
		// save results
        #ifdef _WIN32
            std::regex reg(R"(\.(\w*)$)");
            save_path = std::regex_replace(save_path, reg, ".txt");
            std::ofstream outputFile(save_path);
            if (!outputFile.is_open()) {
                std::cout << "Create txt file fail." << std::endl;
            }
			outputFile << max_index << "\n";
            outputFile.close();
        #elif __linux__
            std::regex reg(R"(\.(\w*)$)");
            save_path = std::regex_replace(save_path, reg, "_result.txt");
            std::ofstream outputFile(save_path);
            if (!outputFile.is_open()) {
                std::cout << "Create txt file fail." << std::endl;
            }
			outputFile << max_index << "\n";
            outputFile.close();
        #endif
    }
		
	}
	//关闭设备
	Device::Close(device);
    return 0;
}

File: mdz/pytorch/PointNet/3_deploy/modelzoo/PointNet/pyrt/PointNet.py
"""
@Description: Main script for testing PointNet accuracy on ModelNet40 dataset
"""

# %% Import packages
import numpy as np
import torch
import sys
import yaml
import os 
import logging
level = logging.INFO
logging.basicConfig(level=level, format='%(asctime)s - [%(levelname)s] - %(message)s')
# Import Icraft packages
import icraft
from icraft.xir import *
from icraft.xrt import *
from icraft.host_backend import *
from icraft.buyibackend import *
# Import pyrt dependencies
sys.path.append(R"../../../Deps/modelzoo")
from pyrtutils.Netinfo import Netinfo
from pyrtutils.utils import *
from pyrtutils.et_device import *
from pyrtutils.icraft_utils import *

# Import preprocess post_process utils
from pointnet_utils import *

def main(config_path) -> None:
    # %% set configurations
    # ---------------------------------参数设置---------------------------------
     # 从yaml里读入配置
    cfg = yaml.load(open(config_path, "r"), Loader=yaml.FullLoader)
    folderPath = cfg["imodel"]["dir"]
    stage = cfg["imodel"]["stage"]
    run_sim = cfg["imodel"]["sim"]
    JSON_PATH, RAW_PATH = getJrPath(folderPath,stage,run_sim)
    
    load_mmu = cfg["imodel"]["mmu"]
    load_speedmode = cfg["imodel"]["speedmode"]
    load_compressFtmp = cfg["imodel"]["compressFtmp"]
    ip = str(cfg["imodel"]["ip"])
    save = cfg["imodel"]["save"]
    show = cfg["imodel"]["show"]

    imgRoot = os.path.abspath(cfg["dataset"]["dir"])
    imgList = os.path.abspath(cfg["dataset"]["list"])
    names_path = cfg["dataset"]["names"]
    resRoot = cfg["dataset"]["res"]
    metricRes = cfg["imodel"]['metricRes']
    #模型自身相关参数配置
    number_of_class = cfg["param"]["number_of_class"]
    
    
    
    # 加载network
    network = loadNetwork(JSON_PATH, RAW_PATH)
    print('INFO: Load network!')
    # 初始化netinfo
    netinfo = Netinfo(network)
    # 选择对网络进行切分
    network_view = network.view(netinfo.inp_shape_opid + 1)
    # 打开device
    device = openDevice(run_sim, ip, netinfo.mmu or load_mmu)
    print('INFO: Open Device!')
    # 初始化session
    session = initSession(run_sim, network_view, device, netinfo.mmu or load_mmu, load_speedmode, load_compressFtmp)
    #开启计时功能
    session.enableTimeProfile(True)
	#session执行前必须进行apply部署操作
    session.apply()
    print('INFO: Session Apply!')

    logging.info(
        f"DATA INFO:\n"
        f"\tData Root Path:                {imgRoot}\n"
        "\n"

        f"Selected Computing Device:\t       {'cpu'}\n"
        "\n"
        f"Simulation Mode:\t                 {run_sim}\n"
        "\n"

        f"TEST INFO:\n"
        f"\tTested Model Path:            {JSON_PATH}\n"
        "\n"
    )
    # %% what kinds of classes do we have?
    classes = CLASS
    logging.info(f"Classes: {classes}")

    # %% data loaders
    logging.info("Loading the data...")
    test_loader = get_dataloader(
        data_path=imgRoot,
        folder="test",
        dataset_type="test",
        batch_size=1,
        num_workers=0,
        pin_memory=True,
        shuffle=False
    )
    # %% let's test the trained network
    logging.info("Testing started...")
    all_preds = []
    all_labels = []
    correct_sum = 0
    if save:
        os.makedirs(resRoot, exist_ok=True)
    for idx, data in enumerate(test_loader):
        inputs = data['pointcloud'].float() #[1,1024,3]
        labels = data['category']
        _inputs = inputs.transpose(1,2).numpy()
        _inputs = np.ascontiguousarray(_inputs)
        # ndarray to icraft.Tensor
        input_tensor = Tensor(_inputs, Layout("**C"))
        # 模型前向推理
        generated_output = session.forward([input_tensor])
        # 前向完，重置device
        if not run_sim: 
            device.reset(1)
        # check outputs
        # [1,40] : pred_results  用(outputs[0])
        # [3,3]  : transform_matrix_3x3
        # [64,64]: transform_matrix_64x64
        # for i in range(3):
        #     out = np.array(generated_output[i])
        #     print(out.shape)
        # print('INFO: get forward results!')
        # 组装成检测结果
        outputs = []
        for i in range(3):
            out = np.array(generated_output[i])
            # print(out.shape)
            outputs.append(torch.tensor(out))
        # 后处理
        _, pred = torch.max(outputs[0], 1) # value,index=pred
        
        print(f'TEST RESULT:\n\tpred =',pred.cpu().numpy() == labels.cpu().numpy())
        # 精度计算
        all_preds += list(pred.cpu().numpy())
        all_labels += list(labels.cpu().numpy())
        correct_sum += (pred.cpu().numpy() == labels.cpu().numpy()).sum()
        # 保存pred结果为txt文件
        np.savetxt(resRoot +'/'+ str(idx)+'_result.txt', pred.cpu().numpy(),fmt="%d")
        pred = pred.cpu().numpy()[0] #np.int64
        for category in CLASS.keys():
            if CLASS[category] == pred:
                break
        
        print(f"\tcategory =",category)
        
    test_acc = correct_sum / len(all_preds) * 100.0
    sResult = f'Top-1: {test_acc}%\n'
    
    # if metricRes, save acc to metric.log
    if metricRes:
        
        # save acc to metric.log
        ACC_PATH = R"./PointNet_metrics.log"
        with open (ACC_PATH,'w') as f:
            f.write(sResult)
            f.close() 
    print('Top-1: %5.1f %% \nPred results save at %s'% ( test_acc,resRoot))  
    logging.info("Testing DONE!")
    # 关闭device
    if not run_sim: 
        Device.Close(device) 
   

if __name__ == '__main__':
    # YAML_CONFIG_PATH = R'../cfg/pointnet_test.yaml'
    Yaml_Path = sys.argv[1]
    # Yaml_Path = "../cfg/pointnet_test_tt.yaml"
    if len(sys.argv) < 2:
        print("Info:未传入yaml参数,读入默认yaml文件:‌{}进行相关配置.".format(Yaml_Path), VERBOSE, 0)        
    if len(sys.argv) == 2:
        Yaml_Path = sys.argv[1]
        print("info:传入yaml文件:‌{}进行相关配置.".format(Yaml_Path), VERBOSE, 0)        
    if len(sys.argv) > 2:
        print("info:传入参数数量错误,请检查运行命令!", VERBOSE, 0)        
        sys.exit(1)
    main(Yaml_Path)
